earth_enterprise/src/scons/khEnvironment.py

```diff
@@ -159,6 +162,192 @@ def EmitBuildDateStrfunc(target, build_date):
+  return 'EmitLongVersion(%s, %s)' % (target, backupFile)
+  
+
+def GetLongVersion(backupFile):
+    """Create a detailed version string based on the state of
+       the software, as it exists in the repository."""
+    
+    if CheckGitAvailable():
+        return GitGeneratedLongVersion()
+
+    # Without git, must use the backup file to create a string.
+    base = ReadBackupVersionFile(backupFile)
+    date = datetime.utcnow().strftime("%Y%m%d%H%M")
+    
+    return '-'.join([base, date])
+
+
+def GetVersion(backupFile):
+    """As getLongVersion(), but only return the leading *.*.* value."""
+    try:
+        raw = GetLongVersion(backupFile)
+        final = raw.split("-")[0]
+    except Exception:
+        return "Version Error"
+
+    return final
+  
+
+def CheckGitAvailable():
+    """Try the most basic of git commands, to see if there is
+       currently any access to a repository."""
+    
+    try:
+        subprocess.check_output(['git', 'status'])
+    except Exception:
+        return False
+    
+    return True
+
+
+def CheckDirtyRepository():
+    """Check to see if the repository is not in a cleanly committed state."""
+    
+    str = subprocess.check_output(['git', 'status', '--porcelain'])
+    
+    # Ignore version.txt for this purpose, as a build may modify the file
+    # and lead to an erroneous interpretation on repeated consecutive builds.
+    if (str == " M earth_enterprise/src/version.txt\n"):
+        return False
+    
+    return (len(str) > 0)
+    
+
+def ReadBackupVersionFile(target):
+    """There should be a file checked in with the latest version
+       information available; if git isn't available to provide
+       information, then use this file instead."""
+    
+    try:
+        fp = open(target, 'r')
+        line = fp.readline()
+        fp.close()
+        
+        return line
+    
+    except Exception:
+        return "Version Error"
+
+
+def GitGeneratedLongVersion():
+    """Take the raw information parsed by git, and use it to
+       generate an appropriate version string for GEE."""
+    try:
+        raw = subprocess.check_output(['git', 'describe', '--tags',
+                                        '--match', '[0-9]*\.[0-9]*\.[0-9]*\-*'])
+        raw = raw.rstrip()
+    except Exception:
+        return "Version Error"
+
+    # Grab the datestamp.
+    date = datetime.utcnow().strftime("%Y%m%d%H%M")
+
+    # If this condition hits, then we are currently on a tagged commit.
+    if (len(raw.split("-")) < 4):
+        if CheckDirtyRepository():
+            return '.'.join([raw, date])
+        return raw
+
+    # Tear apart the information in the version string.
+    components = ParseRawVersionString(raw)
+  
+    # Determine how to update, since we are *not* on tagged commit.
+    if components['isFinal']:
+        components['patch'] = 0
+        components['patchType'] = "alpha"
+        components['revision'] = components['revision'] + 1
```

Comment:
Can you explain how we know that the `revision` part of the version should be increased?  How do we know we shouldn't increase the `major` or `minor` version number instead?

Reply:
There were no rules in the Version documentation for changing the major/minor version numbers. Those changes should come from someone tagging the repository to a newer version.

Reply:
I'm confused.  Why is `revision` increased for a `final` version, and why does it become `alpha`?

Reply:
When the repo is tagged as something like "5.2.1-2.final" (or "5.2.1-2.release", now), that's indicating "we declare 5.2.1 to be done, and this is something that can be released to the public".

If we've started making commits and changes from that point, then we have, by definition, moved on past 5.2.1, and to 5.2.2 (operating under the scenario where there's no clear indicator that we should be moving on to 5.3 instead - I believe someone should tag a "5.3.0-0.beta" commit in order to make that transition, lacking any better ideas).

Reply:
I see. It doesn't seem like a good idea to be making commits towards an unknown version.  I may prefer an error, but I'm OK with this.